File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <bits/stdc++.h>
using namespace std;
const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};
char a[24][24];
int r, c, ret = -987654321, visited[30];
void go(int y, int x, int cnt){
    ret = max(ret, cnt);
    for(int i = 0; i < 4; i++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || ny >= r || nx < 0 || nx >= c)continue;

        int next = a[ny][nx] - 'A';
        if(visited[next] == 0){
            visited[next] = 1;
            go(ny, nx, cnt + 1);
            visited[next] = 0;
        }
    }
}
int main(){
    cin >> r >> c;
    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
            cin >> a[i][j];
        }
    }
    visited[a[0][0] - 'A'] = 1;
    go(0, 0, 1);
    cout << ret << '\n';
    return 0;
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.StringTokenizer;

public class Main{
    static final int[] dx = {-1, 0, 1, 0};
    static final int[] dy = {0, 1, 0, -1};

    static StringBuilder sb = new StringBuilder();
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int r, c;
    static int[][] a;
    static boolean[] visited;
    static int ret;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        r = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());

        a = new int[r][c];
        visited = new boolean[26];

        for(int i = 0; i < r; i++){
            String s = br.readLine();
            for(int j = 0; j < c; j++){
                a[i][j] = s.charAt(j) - 'A'; // 0에서 25까지...
            }
        }

        visited[a[0][0]] = true;
        go(0, 0, 1);

        sb.append(ret);

        bw.write(ret + "");
        bw.flush();
        bw.close();
    }

    static void go(int x, int y, int cnt){
        ret = Math.max(ret, cnt);

        for(int i = 0; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];

            if(nx < 0 || nx >= r || ny < 0 || ny >= c)continue;

            int next = a[nx][ny];
            if(!visited[next]){
                visited[next] = true;
                go(nx, ny, cnt + 1);
                visited[next] = false;
            }
        }
    }
}